oeps/oep-0006.rst

```diff
@@ -0,0 +1,224 @@
+* ``Scope.preferences`` (UserScope.ONE, BlockScope.TYPE)
+
+  * Example: (For a video XBlock) Boolean setting to prefer YouTube or non-YouTube video playback.
+
+
+Motivation
+==========
+
+As one would expect, ``Scope.user_info`` and ``Scope.preferences`` fields in the Open edX platform are implemented globally, and not scoped to the course that uses them. There is currently no standard, extensible way to use XBlock fields to store data that is scoped to a particular "context," such as a course, although this is a common need.
+
+
+Specification
+=============
+
+New BlockScope values
+---------------------
+
+To support context-scoped fields, ``BlockScope`` will allow these new additional values:
+
+* ``CONTEXT_TYPE``: The data is related to all instances of this type of XBlock within a particular context.
+* ``CONTEXT_ALL``: The data is accessible to XBlocks of any type, within a particular context. (As with ``BlockScope.ALL``, XBlock authors who use such fields should be careful to use a namespace that will not conflict with ``BlockScope.CONTEXT_ALL`` fields defined by other XBlock authors.)
+
+A "context" is defined by the runtime but is usually a "course" or the nearest equivalent. In the Open edX platform, the context may be a course or a content library. Different runs of the same course would be considered different contexts and would not share data, though UserScope.NONE fields would be copied when creating a new course run. On other platforms, a "context" could be defined as a blog post or a textbook.
+
+New named scopes
+----------------
+
+As is the situtation today, XBlock runtimes would generally not be expected to implement every possible combination of ``BlockScope`` and ``UserScope``. Instead, XBlock runtimes are encouraged to implement these new named scopes, which would be defined in the XBlock python package:
+
+* ``Scope.context_settings`` (UserScope.NONE, BlockScope.CONTEXT_TYPE)
+
+  * Example: For an LTI XBlock, a list of the LTI provider URLs and secrets used in the course
+  * Example: For a discussions XBlock: Enable or disable discussions in this course
+
+* ``Scope.context_shared_settings`` (UserScope.NONE, BlockScope.CONTEXT_ALL)
+
+  * Example: Course-specific defaults that can apply to many different blocks, such as "max_attempts" and "show_answer", which would be consumed by the Capa Problem XBlock, Drag and Drop problem XBlock, Problem Builder XBlock family, etc. (These are currently implemented via ``InheritanceMixin``, which is built-in to edx-platform and does not allow XBlock plugins to extend the list of inherited fields. However, a downside of making them a context-scoped field is that each block would require two field definitions: a ``Scope.settings`` field for the usage-specific value, and a ``Scope.context_shared_settings`` field for the default/fallback value. )
+  * Example: Submission blackout dates - don't allow users to submit answers during the specified time period.
```

Comment:
@robrap Sure. I have already removed `Scope.context_shared_preferences` from the proposal. I think for now I will remove `Scope.context_shared_settings` as well. Per your suggestion, I'll put them into "Rejected Alternatives" or similar.